subsys/bluetooth/host/conn.c

```diff
@@ -1205,45 +1220,43 @@ static void notify_tx(void)
 	}
 }
 
-static sys_snode_t *add_pending_tx(struct bt_conn *conn, bt_conn_tx_cb_t cb,
-				   void *user_data)
+static struct bt_conn_tx *add_pending_tx(struct bt_conn *conn,
+					 bt_conn_tx_cb_t cb, void *user_data)
 {
 	struct bt_conn_tx *tx;
-	sys_snode_t *node;
 	unsigned int key;
 
 	BT_DBG("conn %p cb %p user_data %p", conn, cb, user_data);
 
-	__ASSERT(!sys_slist_is_empty(&free_tx), "No free conn TX contexts");
-
-	node = sys_slist_get_not_empty(&free_tx);
-	tx = CONTAINER_OF(node, struct bt_conn_tx, node);
+	tx = k_fifo_get(&free_tx, K_FOREVER);
+	tx->conn = conn;
```

Comment:
This should actually be `tx->conn = bt_conn_ref(conn)` and then the corresponding `unref` in the `tx_free()` function

Reply:
Would this prevent the conn to be free if that are still refs to the conn, I though that could prevent conn_cleanup to be run or that just depend on the DISCONNECTED state?

Reply:
You shouldn't confuse the connection state with connection reference counting. Those are completely independent things. The reference counting is simply counting how many pointers variables you have pointing at the same conn object, regardless of what the state is.